module2/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,149 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.dao.GiftCertificateDao;
+import com.epam.esm.dao.entity.GiftCertificate;
+import com.epam.esm.dao.entity.Tag;
+import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.service.TagService;
+import com.epam.esm.dao.util.DynamicQueryResult;
+import com.epam.esm.service.dto.GiftCertificateDto;
+import com.epam.esm.dao.util.GiftCertificateParameter;
+import com.epam.esm.service.dto.TagDto;
+import com.epam.esm.service.exception.ServiceException;
+import com.epam.esm.service.mapper.GiftCertificateMapper;
+import com.epam.esm.service.mapper.TagMapper;
+import com.epam.esm.dao.util.DynamicQuery;
+import com.epam.esm.service.validator.Validator;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.epam.esm.service.util.ErrorMessage.ERROR_GIFT_CERTIFICATE_CREATE;
+import static com.epam.esm.service.util.ErrorMessage.ERROR_GIFT_CERTIFICATE_CREATE_DTO;
+import static com.epam.esm.service.util.ErrorMessage.ERROR_GIFT_CERTIFICATE_GET_BY_ID;
+import static com.epam.esm.service.util.ErrorMessage.ERROR_GIFT_CERTIFICATE_GET_UPDATE;
+import static com.epam.esm.service.util.ErrorMessage.ERROR_GIFT_CERTIFICATE_GET_UPDATE_DTO;
+import static com.epam.esm.service.util.ErrorMessage.ERROR_PARAM;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+    private final GiftCertificateDao giftCertificateDao;
+    private final TagService tagService;
+    private final GiftCertificateMapper giftCertificateMapper;
+    private final TagMapper tagMapper;
+    private final Validator validator;
+
+    @Override
+    @Transactional
+    public GiftCertificateDto create(GiftCertificateDto giftCertificateDto) {
+        if (!validator.validatedGiftCertificateDto(giftCertificateDto)) {
+            throw new ServiceException(ERROR_GIFT_CERTIFICATE_CREATE_DTO);
```

Comment:
It's better to return 400 error code (Bad request) for validation errors, but you return 404 always

Reply:
fixed